### ISSUE TITLE
Auto-generate .ensime from Emacs (or equivalent)

### ISSUE THREAD

**Issue by eallik** [ensime/ensime-server#773]
Eclipse, IntelliJ and other IDEs typically don't need an external re-generation of the project file after changes to dependencies or other project settings, whereas it seems that with Ensime, a re-run of `gen-ensime` on the SBT console is needed after most changes to `build.sbt`.

Is there a way to somehow make this transparent/automatic? Also, are there any plans to make Ensime automatically verify the consistency and up-to-date-ness of `.ensime_cache` so that after updating Ensime, it wouldn't be (generally always) necessary to `rm -rf .ensime_cache`?

**Comment by fommil** [comment]
errr, you should never need to `rm -rf .ensime_cache` !! If you can tell us when you need to do that, it would be very interesting.

With regards to refreshing `gen-ensime`, you could write a lisp script to detect timestamp changes and then re-run `sbt gen-ensime` followed by a server restart. This isn't something I'd want to support by default because it will be heavily dependent on user details. I certainly wouldn't want this to happen.

**Comment by edani** [comment]
Regarding running `sbt gen-ensime` automatically, that would be a good thing to add to the doc. @eallik if you can figure it out it would be appreciated. Otherwise I can look into it.

**Comment by edani** [comment]
Also on this topic, it would be nice if ENSIME server could support updating the project classpath, and maybe other aspects of the config, without a server restart. It might be just a matter of restarting the indexer.

**Comment by fommil** [comment]
I don't know if it's worth it... once the dependency jars are in place, the server starts up from cold basically instantly, and the indexer is designed to minimise the changes it needs to make to an existing persistent database. That's why deleting the cache is the worst thing to do, performance wise.

**Comment by fommil** [comment]
:+1: to an emacs function that can do the reload.

**Comment by aemoncannon** [comment]
+1 for having an immutable config..if we can get away with it.

**Comment by codygman** [comment]
I came here to recommend automating generating the .ensime file as well. I know it's "just typing in ` sbt gen-ensime`" for every new project, but I work with 10+ languages and it's nice to reduce all those "you just have to do X simple thing(s)" as much as possible.

Then you end up with interactive environments across languages where you don't have to remember a lot of (usually meaningless for the task) plumbing.

**Comment by fommil** [comment]
Running the generate task can be very slow, and there is no easy way to auto detect a change in the build, so that's why we've not done this yet. Open to suggestions, and especially PRs :-)

**Comment by eallik** [comment]
but how do Eclipse and IDEA do it?

**Comment by fommil** [comment]
Most people I know turn off that "auto refresh" feature because it locks up the IDE. Refreshing from an SBT project can take ~10 minutes on corporate sized projects thanks to awesome ivy resolution (even when the files are held locally).

We could add support for trivial projects that only have a `build.sbt` (in which case it's as simple as comparing the timestamps of `.ensime` and `build.sbt`), or move to using `sbt-server` as in #849 but that's for re-generating, not generating. We could have hacky support that checks the timestamp of all `*.sbt` files in the root and all `.scala` files under the `project` directory... that'll probably work for most people.

As for generating the file the first time, this could be easier ... detecting an sbt project is relatively easy. But this is then specific sbt logic and not everybody is using sbt.

In either case, I think auto-generation and restarting of the ENSIME server should be **off** by default, although I would be open to asking the user if she wishes to (re-)generate a `.ensime` file when doing `M-x ensime` and the file is not detected or is out of date.

We'd welcome a PR in this area, it's not a priority for the next few releases. As a workaround, you could create your own wrapper around `ensime` that calls out to sbt and runs `gen-ensime` before starting the server.

**Comment by aemoncannon** [comment]
+1
